File: Full Stack Develop/iOS/Objective-C/ObjectiveC Study/ObjectiveC Study/Practice/BuyTickets.m
//
//  BuyTickets.m
//  ObjectiveC Study
//
//  Created by Harry Xiong on 2024/6/21.
//

#import <Foundation/Foundation.h>
#import "BuyTickets.h"

// Aim: Create a ticket agent to help people buy ticket.


@implementation TicketAgent

- (void)buyTicket {
    NSLog(@"Already bought a ticket!");
}

@end

@implementation MyPerson

- (instancetype)initWithTickerAgent:(id<HandleTicket>)agent {
    self = [super init];
    if (self) {
        _delegate = agent;
    }
    return self;
}

- (void)goToWatchMovies {
    NSLog(@"Call TicketAgent to buy a ticket.");
    if ([self.delegate respondsToSelector:@selector(buyTicket)]) {
        [self.delegate buyTicket];
    }
    NSLog(@"I have a ticket. I can go~");
}


@end

void testBuyTick(void) {
    TicketAgent *agent = [TicketAgent new];
    
    MyPerson *person = [[MyPerson alloc] initWithTickerAgent:agent];
    
    [person goToWatchMovies];
}

File: Full Stack Develop/iOS/Swift/Swift Study/Swift Study/Practice/BuyTickets.swift
//
//  BuyTickets.swift
//  Swift Study
//
//  Created by Harry Xiong on 2024/6/24.
//

import Foundation

// Aim: Create a ticket agent to help people buy tickets.

protocol HandleTicket {
    func buyTicket();
}

struct MyPerson {
    
    let agentDelegate: HandleTicket?;
    
    init(agent: HandleTicket? = nil) {
        self.agentDelegate = agent
    }
    
    func goToWatchMovie() {
        print("Called Ticket Agent to buy a ticket.")
        if let delegate = agentDelegate {
            delegate.buyTicket()
        }
        print("I have a ticket. I can go~")
    }
    
}

struct TicketAgent : HandleTicket {
    func buyTicket() {
        print("Already bought a ticket!")
    }
}

func testBuyTicket() {
    let agent = TicketAgent()
    let person = MyPerson(agent: agent);
    
    person.goToWatchMovie();
}
